File: java/Raytracer with shadows and IO/Src/Plane.java
/*
  This class models the xz plane (in a single color)
*/

public class Plane extends Surface
{
  public Plane(RGBColor color)
  {
	 this.color = color;
  }
  
  public boolean hit(Ray ray, double tMin, double tMax, double time, HitRecord record)
  {
    Ray r = getLocalRay(ray);
	 
	 if ( Math.abs(r.direction.body[1]) <= Util.nearlyZero )
	   return false;
		
    double t = - r.origin.body[1] / r.direction.body[1];	 
	 
	 boolean outOfRange = (t < tMin) || (t > tMax);

	 if ( outOfRange  )
      return false;
			 
   record.t = t;
	record.point = r.pointAt(t);
	record.s = this;
	 
   return true;
  }
  
  public boolean shadowHit(Ray ray, double tMin, double tMax, double time)
  {
    Ray r = getLocalRay(ray);	 
	 if ( Math.abs(r.direction.body[1]) <= Util.nearlyZero )
	   return false;
		
    double t = - r.origin.body[1] / r.direction.body[1];	 
	 
	 boolean outOfRange = (t < tMin) || (t > tMax);

    return ( !outOfRange );
	}
  
  public Vector3 normalAt(Vector3 point)
  {
	 return new Vector3(0, 1, 0);
  }
  
  public void dump()
  {
    System.out.println("The xz-plane");
	 System.out.println();
	 System.out.print("Color: ");
	 color.dump();
	 System.out.println();
  }
}

File: src/Tracer/Objects/Plane.cs
using System;
using System.Drawing;
using System.Linq;
using System.Numerics;
using MathNet.Numerics.LinearAlgebra;

namespace Tracer.Objects
{
    public class Plane : Traceable
    {
        public Plane() { }
        public Plane(Color baseColor)
        {
            BaseColor = baseColor;
        }

        public override float[] Intersections(Ray ray)
        {
            var r = GetLocalRay(ray);

            if (Math.Abs(r.Direction.Y) <= Util.NearlyZero)
            {
                return new float[0];
            }

            var t = -r.Origin.Y / r.Direction.Y;

            // out of range

            return new float[] { t };
        }

        public override Vector3 GetNormalAt(Vector3 p)
            => new Vector3(0, 1, 0);
    }
}